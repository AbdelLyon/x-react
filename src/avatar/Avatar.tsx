import type { AvatarProps, UserProps } from "@nextui-org/react";
import { forwardRef } from "react";
import type { AvatarGroupProps } from "@nextui-org/react";
import {
  Avatar as AvatarRoot,
  User as UserRoot,
  useAvatarGroup,
  AvatarGroupProvider,
} from "@nextui-org/react";

export const Avatar = forwardRef<HTMLDivElement, AvatarProps>((props, ref) => {
  return <AvatarRoot ref={ref} {...props} />;
});

Avatar.displayName = "Avatar";

export const AvatarGroup = forwardRef<HTMLDivElement, AvatarGroupProps>(
  (props, ref) => {
    const {
      Component,
      clones,
      context,
      remainingCount,
      renderCount = (count) => <Avatar name={`+${count}`} />,
      getAvatarGroupProps,
    } = useAvatarGroup({
      ref,
      ...props,
    });

    return (
      <Component {...getAvatarGroupProps()}>
        <AvatarGroupProvider value={context}>
          {clones}
          {remainingCount > 0 && renderCount(remainingCount)}
        </AvatarGroupProvider>
      </Component>
    );
  },
);

AvatarGroup.displayName = "AvatarGroup";

export const UserAvatar = forwardRef<HTMLDivElement, UserProps>(
  (props, ref) => {
    return <UserRoot ref={ref} {...props} />;
  },
);

UserAvatar.displayName = "UserAvatar";

Avatar.displayName = "Avatar";
