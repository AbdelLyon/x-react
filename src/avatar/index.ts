import "@/index.css";
export { Avatar } from "@/avatar/Avatar";
