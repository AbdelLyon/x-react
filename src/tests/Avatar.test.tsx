import { Avatar } from "@/avatar";
import { render } from "@testing-library/react";
import type { JSX } from "react";
import { describe, it, expect } from "vitest";

describe("Avatar", () => {
  it("should render default avatar icon when no props are provided", () => {
    const { getByRole } = render(<Avatar />);
    const avatarIcon = getByRole("img");
    expect(avatarIcon).toBeTruthy();
  });

  it("should render image when src is provided", () => {
    const testSrc = "https://example.com/avatar.jpg";
    const { getByRole } = render(<Avatar src={testSrc} alt="Test Avatar" />);

    const image = getByRole("img", { name: "Test Avatar" });
    expect(image).toBeTruthy();
    expect(image.getAttribute("src")).toBe(testSrc);
  });

  it("should render custom fallback component", () => {
    const CustomFallback = (): JSX.Element => <div>Custom Fallback</div>;

    const { getByText } = render(
      <Avatar fallback={<CustomFallback />} showFallback />,
    );

    const fallback = getByText("Custom Fallback");
    expect(fallback).toBeTruthy();
  });

  it("should match snapshot with default props", () => {
    const { asFragment } = render(<Avatar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with image", () => {
    const { asFragment } = render(
      <Avatar src="https://example.com/avatar.jpg" alt="Test Avatar" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with name", () => {
    const { asFragment } = render(<Avatar name="John Doe" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with custom fallback", () => {
    const CustomFallback = (): JSX.Element => <div>Custom Fallback</div>;

    const { asFragment } = render(
      <Avatar fallback={<CustomFallback />} showFallback />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
