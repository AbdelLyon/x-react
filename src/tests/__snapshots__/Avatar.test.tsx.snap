// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > should match snapshot with custom fallback 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <img
      alt="avatar"
      class="flex object-cover w-full h-full transition-opacity !duration-500 opacity-0 data-[loaded=true]:opacity-100"
    />
    <div
      aria-label="avatar"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex items-center justify-center"
      role="img"
    >
      <div>
        Custom Fallback
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <img
      alt="avatar"
      class="flex object-cover w-full h-full transition-opacity !duration-500 opacity-0 data-[loaded=true]:opacity-100"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with image 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <img
      alt="Test Avatar"
      class="flex object-cover w-full h-full transition-opacity !duration-500 opacity-0 data-[loaded=true]:opacity-100"
      src="https://example.com/avatar.jpg"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar > should match snapshot with name 1`] = `
<DocumentFragment>
  <div
    class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
    tabindex="-1"
  >
    <img
      alt="John Doe"
      class="flex object-cover w-full h-full transition-opacity !duration-500 opacity-0 data-[loaded=true]:opacity-100"
    />
  </div>
</DocumentFragment>
`;
